### ISSUE TITLE
Recovering stack traces in presence of inlining

### ISSUE THREAD

**Issue by awelc** [facebook/redex#386]
I have been playing with the `MethodInlinePass` and it seems to bring some benefit to our app, but the problem is that it messes up stack traces in case and exception is raised inside an inlined method.

It would seem that this would also be a problem for FB in gathering stack traces for exceptions raised by the app in the field

It almost looks like there is some support of this in ReDex but it's rather difficult to put all the pieces together. 

In particular, the inliner (Inliner.cpp) seems to be recording some meta data for inlined call sites, but I am not sure how it's being used:
```c++
    // Logging before the call to inline_method to get the most relevant line
    // number near callsite before callsite gets replaced. Should be ok as
    // inline_method does not fail to inline.
    log_opt(INLINED, caller_method, callsite->insn);
```

Also, part of ReDex handling ProGuard mappings seems to have some code for handling inlined methods:
```c++
    // There may be multiple remapped frames if the given instruction was
    // inlined. Create a linked list of DexPositions corresponding to the call
    // chain.
    auto insert_it = code->iterator_to(mie);
    for (auto prev = it++; it != remapped_frames.end(); prev = it++) {
      auto next_pos = std::make_unique<DexPosition>(
          it->method, file_name_from_method_string(it->method), it->line);
      pos->parent = next_pos.get();
      pos = next_pos.get();
      insert_it = code->insert_before(insert_it, std::move(next_pos));
    }
```

Nevertheless, with a very simple config file (below), stack traces reported include incorrect line/file information for the place where the exception has been raised in an inlined method.

```
{
  "redex" : {
    "passes" : [
      "MethodInlinePass",
      "LocalDcePass",
      "RegAllocPass"
    ]
  },
  "MethodInlinePass": {
    "throws": false
  }
}
```

I would be very grateful for any suggestions on how to handle this.

**Comment by justinjhendrick** [comment]
I'm glad MethodInlinePass is useful to you. Yes, you're right, the raw stack traces are incorrect and they need some processing. The metadata you want can be output if you add this to your config at the top level (same level as "redex" or a pass name)

```
"line_number_map_v2": "redex-line-number-map-v2"
```

See [DexPosition.h](https://github.com/facebook/redex/blob/7059429b308795e7111a9c5425371c3aad18cdee/libredex/DexPosition.h) and [DexPosition.cpp](https://github.com/facebook/redex/blob/7059429b308795e7111a9c5425371c3aad18cdee/libredex/DexPosition.cpp#L121) for details of the binary format of the file

**Comment by justinjhendrick** [comment]
I'm not sure exactly how it happens, but I think Redex can replace line numbers with unique identifiers, then the `line_number_map_v2` can convert those identifiers back into filenames and line numbers

**Comment by awelc** [comment]
Thank you for the info!

Is there someone that can provide some additional details? I have been
digging through ReDex's source code but it's tough to find info...

**Comment by justinjhendrick** [comment]
@danzimm has been touching similar areas recently
@int3 worked on the symbolicator a while ago

One or both of them may have some info to add

**Comment by justinjhendrick** [comment]
I think [symbolicate-trace.cpp](https://github.com/facebook/redex/blob/916eac2afdec31f84de5d99f083aa38a3e4fa29b/tools/debug-info/symbolicate-trace.cpp) can symbolicate stack traces given a rename map, like this:

```
$ cat stacktrace.txt | remap redex_line_number_map_v2
```

**Comment by awelc** [comment]
Thank you for additional details! The remaining question, I guess, is how to make ReDex replace line numbers with identifiers. I hope this part is also somewhere in the open source version...

**Comment by justinjhendrick** [comment]
I was wondering that too. I think `"line_number_map_v2": "redex-line-number-map-v2"` might also enable the use of unique identifiers instead of regular line numbers?

**Comment by awelc** [comment]
Indeed this config change both generates the map file and replaces line numbers for inlined methods with unique id-s. I think I got all I needed! Really appreciate the help!!!

**Comment by justinjhendrick** [comment]
Seems resolved. Feel free to re-open for a follow up question though.

**Comment by danzimm** [comment]
Hey @awelc, I'm really sorry about the delay getting back to this thread. The last two weeks have been pretty hectic at work.

It seems like you've been able to get your questions answered. As @justinjhendrick mentioned, please don't hesitate to re-open the issue if you have any other questions!

**Comment by int3** [comment]
Hey @awelc -- just a heads up that I've landed ec48b3c, which changes the config file API for debug info. To keep the behavior you currently have, you'll want to change `debug_info_kind` in `config/default.config` from `no_custom_symbolication` to `per_method_debug`. The `line_number_map_v2` entry is no longer necessary (the symbol file will always be named `redex-line-number-map-v2`).

**Comment by danzimm** [comment]
Hey @awelc sorry for bugging you again. I wanted to let you know that I landed a few commits that ends in https://github.com/facebook/redex/commit/d23403802f576f5a1e0139ac810f781be16ddc29 which together open source a python tool we use to symbolicate traces pulled from optimized builds.

I hope this helps you out! Let me know if I can provide any support for the symbolication lib/script!
